server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java

```diff
@@ -221,18 +254,27 @@ public FlatObjectFieldType(String name, NamedAnalyzer analyzer) {
             this.mappedFieldTypeName = null;
         }
 
-        public FlatObjectFieldType(String name, String mappedFieldTypeName) {
+        public FlatObjectFieldType(
+            String name,
+            String mappedFieldTypeName,
+            KeywordFieldMapper.KeywordFieldType valueFieldType,
+            KeywordFieldMapper.KeywordFieldType valueAndPathFieldType,
+            boolean isSearchable,
+            boolean hasDocValues
+        ) {
             super(
                 name,
-                true,
+                isSearchable,
                 false,
-                true,
+                hasDocValues,
                 new TextSearchInfo(Defaults.FIELD_TYPE, null, Lucene.KEYWORD_ANALYZER, Lucene.KEYWORD_ANALYZER),
                 Collections.emptyMap()
             );
             this.ignoreAbove = Integer.MAX_VALUE;
             this.nullValue = null;
             this.mappedFieldTypeName = mappedFieldTypeName;
+            this.valueFieldType = valueFieldType;
```

Comment:
The way the `valueFieldType` and `alueAndPathFieldType` are assigned is through setters, we should stick to the same strategy everywhere: use either setters or constructors. Otherwise it is very hard to trace what the exact values will be.